soroban-env-host/src/events/internal.rs

```diff
@@ -156,30 +170,45 @@ impl InternalEventsBuffer {
     /// Converts the internal events into their external representation. This should only be called
     /// either when the host is finished (via `try_finish`), or when an error occurs.
     pub fn externalize(&self, host: &Host) -> Result<Events, HostError> {
-        let vec: Result<Vec<HostEvent>, HostError> = self
-            .vec
-            .iter()
-            .map(|e| match &e.0 {
+        let mut vec: Vec<HostEvent> = vec![];
+        // This line is intentionally unmetered, because metering is taken care during event push
+        // to ensure metering is indifferent of the buffer length.
+        vec.reserve(self.vec.len());
+
+        for (event, status) in self.vec.iter() {
+            match event {
                 InternalEvent::Contract(c) => {
                     // Metering: we use the cost of instantiating a size=1 `Vec` as an estimate
                     // for the cost collecting 1 `HostEvent` into the events buffer. Because
                     // the resulting buffer length may be different on different instances
                     // (due to diagnostic events) and we need a deterministic cost across all
                     // instances, the cost needs to be amortized and buffer size-independent.
                     Vec::<HostEvent>::charge_bulk_init_cpy(1, host)?;
-                    Ok(HostEvent {
+                    vec.push(HostEvent {
                         event: c.to_xdr(host)?,
-                        failed_call: e.1 == EventError::FromFailedCall,
-                    })
+                        failed_call: *status == EventError::FromFailedCall,
+                    });
                 }
-                InternalEvent::Diagnostic(c) => host.as_budget().with_free_budget(|| {
-                    Ok(HostEvent {
-                        event: c.to_xdr(host)?,
-                        failed_call: e.1 == EventError::FromFailedCall,
-                    })
-                }),
-            })
-            .collect();
-        Ok(Events(vec?))
+                InternalEvent::Diagnostic(d) => {
+                    // If the host is not in debug mode, the diagnostic event
+                    // (which should not be generated in the first place), will be incored.
+                    let de: Option<HostEvent> = host.with_debug_budget(
+                        || match d.to_xdr(host)? {
+                            Some(event) => Ok(Some(HostEvent {
+                                event,
+                                failed_call: *status == EventError::FromFailedCall,
+                            })),
+                            None => Ok(None),
+                        },
+                        || None,
+                    );
+                    if let Some(de) = de {
+                        vec.push(de)
```

Comment:
This should also do `charge_bulk_init_cpy` like the observable Contract-event case. It's just charging to the debug budget.

Reply:
You are right, this part is rewritten (with this fix and ones mentioned below).